graphistry/collapse.py

```diff
@@ -0,0 +1,532 @@
+import pandas as pd
+import numpy as np
+
+COLLAPSE_NODE = "collapse_node"
+COLLAPSE_SRC = "collapse_src"
+COLLAPSE_DST = "collapse_dst"
+WRAP = "~"
+DEFAULT_VAL = "None"
+VERBOSE = True
+
+
+def unpack(g):
+    """
+        Helper method that unpacks graphistry instance
+    ex:
+        ndf, edf, src, dst, node = unpack(g)
+
+    -----------------------------------------------------------------------------------------
+
+    :param g: graphistry instance
+    :returns node DataFrame, edge DataFrame, source column, destination column, node column
+    """
+    ndf = g._nodes
+    edf = g._edges
+    src = g._source
+    dst = g._destination
+    node = g._node
+    return ndf, edf, src, dst, node
+
+
+def get_children(g, node_id, hops=1):
+    """
+        Helper that gets children at k-hops from node `node_id`
+
+    ------------------------------------------------------------------
+
+    :returns
+    """
+    g2 = g.hop(pd.DataFrame({g._node: [node_id]}), hops=hops)
+    return g2
+
+
+def has_edge(g, n1, n2, directed=True) -> bool:
+    """
+        Checks if `n1` and `n2` share an (directed or not) edge
+
+    ------------------------------------------------------------------
+
+    :param g: graphistry instance
+    :param n1: `node` to check if has edge to `n2`
+    :param n2: `node` to check if has edge to `n1`
+    :param directed: bool, if True, checks only outgoing edges from `n1`->`n2`, else finds undirected edges
+    :returns bool, if edge exists between `n1` and `n2`
+
+    """
+    ndf, edf, src, dst, node = unpack(g)
+    if directed:
+        if n2 in edf[edf[src] == n1][dst].values:
+            return True
+    else:
+        if (n2 in edf[edf[src] == n1][dst].values) or (
+            n1 in edf[edf[src] == n2][dst].values
+        ):
+            return True
+    return False
+
+
+def get_edges_of_node(g, node_id, directed=True, hops=1):
+    """
+        Gets edges of node at k-hops from node
+
+    ----------------------------------------------------------------------------------
+
+    :param g: graphistry instance
+    :param node_id: `node` to find edges from
+    :param directed: bool, if true, finds all outgoing edges of `node`, default True
+    :param hops: the number of hops from `node` to take, default = 1
+    :returns DataFrame of edges
+    """
+    _, _, src, dst, _ = unpack(g)
+    g2 = get_children(g, node_id, hops=hops)
+    if directed:
+        edges = g2._edges[dst]
```

Comment:
need a `.drop_duplicates()` here too... move to be part of the return?

Reply:
4 eyes are better than 2, thanks